diff --git a/code/tune2.py b/code/tune2.py
index 68479a1..d023f54 100644
--- a/code/tune2.py
+++ b/code/tune2.py
@@ -1,5 +1,5 @@
 import os
-cuda = 1
+cuda = 0
 # Baseline--yelp2018--2 LightGCN_PyG + BPR
 os.system(f'python main.py --project GCLRec_No_Valid --name LightGCN_PyG+BPR --model LightGCN_PyG --loss BPR --dataset yelp2018\
             --init_method Normal --lr 0.001 --weight_decay 1e-4 --cuda {cuda} --num_layers 3 --latent_dim_rec 64 --batch_size 2048\
